pkg/k8s/apis/cilium.io/v2alpha1/bgp_node_types.go

```diff
@@ -0,0 +1,183 @@
+// Copyright Authors of Cilium
+
+package v2alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:categories={cilium,ciliumbgp},singular="ciliumbgpnodeconfig",path="ciliumbgpnodeconfigs",scope="Cluster",shortName={bgpnode}
+// +kubebuilder:printcolumn:JSONPath=".metadata.creationTimestamp",name="Age",type=date
+// +kubebuilder:storageversion
+
+// CiliumBGPNodeConfig is node local configuration for BGP agent. Name of the object should be node name.
+// This resource will be created by Cilium operator and is read-only for the users.
+type CiliumBGPNodeConfig struct {
+	// +deepequal-gen=false
+	metav1.TypeMeta `json:",inline"`
+	// +deepequal-gen=false
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec is the specification of the desired behavior of the CiliumBGPNodeConfig.
+	Spec CiliumBGPNodeSpec `json:"spec"`
+
+	// Status is the most recently observed status of the CiliumBGPNodeConfig.
+	Status CiliumBGPNodeStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=false
+// +deepequal-gen=false
+
+// CiliumBGPNodeConfigList is a list of CiliumBGPNodeConfig objects.
+type CiliumBGPNodeConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	// Items is a list of CiliumBGPNodeConfig.
+	Items []CiliumBGPNodeConfig `json:"items"`
+}
+
+type CiliumBGPNodeSpec struct {
+	// BGPInstances is a list of BGP router instances on the node.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +listType=map
+	// +listMapKey=name
+	BGPInstances []CiliumBGPNodeInstance `json:"bgpInstances"`
+}
+
+// CiliumBGPNodeInstance is a single BGP router instance configuration on the node.
+type CiliumBGPNodeInstance struct {
+	// Name is the name of the BGP instance.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// LocalASN is the ASN of this virtual router.
+	// Supports extended 32bit ASNs.
+	//
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=4294967295
+	LocalASN *int64 `json:"localASN,omitempty"`
+
+	// RouterID is the BGP router ID of this virtual router.
+	// This configuration is derived from CiliumBGPNodeConfigOverride resource.
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Format=ipv4
+	RouterID *string `json:"routerID,omitempty"`
+
+	// LocalPort is the port on which the BGP daemon listens for incoming connections.
+	//
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	// +kubebuilder:default=179
+	LocalPort *int32 `json:"localPort,omitempty"`
+
+	// Peers is a list of neighboring BGP peers for this virtual router
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
```

Comment:
nit: Not a bad idea to establish a max number of bgp peers and set `// +kubebuilder:validation:MaxItems`.

Reply:
This is a bit tricky, as the number of peers can differ from deployment to deployment. I am leaning towards not setting any max on number of peers. 

Also, made this field optional.